pallets/settlement/src/lib.rs

```diff
@@ -1522,55 +1524,101 @@ impl<T: Trait> Module<T> {
                 receipt.metadata.clone(),
             ));
         }
-
-        // Execute instruction if conditions are met.
-        let execute_instruction_weight = Self::is_instruction_executed(
-            affirms_pending,
-            instruction_details.settlement_type,
-            instruction_id,
-        );
-
-        execute_instruction_weight.map(|info| {
-            info.actual_weight
-                .map(|w| {
-                    w.saturating_add(weight_for::weight_for_affirmation_with_receipts::<T>(
-                        u32::try_from(receipt_details.len()).unwrap_or_default(),
-                    ))
-                })
-                .into()
-        })
+        Ok(())
     }
 
     pub fn base_affirm_instruction(
         origin: <T as frame_system::Trait>::Origin,
         instruction_id: u64,
         portfolios: Vec<PortfolioId>,
-    ) -> DispatchResultWithPostInfo {
+    ) -> DispatchResult {
         let (did, secondary_key) =
             Self::ensure_origin_perm_and_instruction_validity(origin, instruction_id)?;
         let portfolios_set = portfolios.into_iter().collect::<BTreeSet<_>>();
 
         // Provide affirmation to the instruction
-        Self::unsafe_affirm_instruction(
-            did,
+        Self::unsafe_affirm_instruction(did, instruction_id, portfolios_set, secondary_key.as_ref())
+    }
+
+    // It affirms the instruction and may schedule the instruction
+    // depends on the settlement type.
+    pub fn affirm_with_receipts_and_may_schedule_instruction(
+        origin: <T as frame_system::Trait>::Origin,
+        instruction_id: u64,
+        receipt_details: Vec<ReceiptDetails<T::AccountId, T::OffChainSignature>>,
+        portfolios: Vec<PortfolioId>,
+    ) -> DispatchResult {
+        Self::base_affirm_with_receipts(origin, instruction_id, receipt_details, portfolios)?;
+        // Schedule instruction to be execute in the next block (expected) if conditions are met.
+        Self::maybe_schedule_instruction(
```

Comment:
For simplicity, I'd suggest adding `maybe_schedule_instruction` to the native `base_affirm*` functions so that we don't have to worry about which variant we have to call. It only costs 1 read to verify if the instruction needs to be scheduled so it does not carry a big cost.

Reply:
I really don't want to make a function to complex to understand so I think it is good to have segregation and it will improve code readability.

Reply:
I think the current change reduces code readability. The reader has to keep track of when the proposal is being only approved and when it might be scheduled. In reality, they don't need to do that. The same logic is applicable to both scenarios.

Reply:
I don't think so, right now I am using with_transaction() for the STO case while we don't need `with_transaction` for the scheduling case.

Reply:
I am not talking about `affirm_and_execute_instruction`. That's a completely different function and it's fine to have that.

I am talking about `affirm_and_maybe_schedule_instruction` vs `base_affirm_instruction` and similar functions. Basically, I think `affirm_with_receipts_and_maybe_schedule_instruction` and similar functions should not exist.